### PR TITLE
[Feature: InstructorUI] Allow supplying only user_id for uploading user (grader / class) list

### LINKED ISSUE #4917: Allow supplying only user_id for uploading grader list
**What problem are you trying to solve with Submitty**
When uploading the grader list via CSV (or XLSX), it is required to have the entire row filled out, which can be cumbersome to collect. It would be nice, much like the user add box, to just supply a user_id and have the other fields auto-populated.

### PR DESCRIPTION
**Pull request by mukul-kmr-jha**
### Please check if the PR fulfills these requirements:

* [x] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)

### What is the current behavior?
Currently, It is required to fill at least 7 columns (pref first and last name being the optional) for each row when uploading the graders' list. 

### What is the new behavior?
Closes #4917 
Now You have the option to just specify the user_id in the file to let auto-population fill other details. 
### Other information?
When using auto-population, the 'limited access grader' privilege is given to the user by default.

### CODE REVIEW

**Comment by BenKelly99** [comment]
image
Two things:
1. Move up the "starred" part to be with the explanation of the csv format as having it at the bottom is a little confusing.
2. The explanation is confusing to someone who doesn't know what the auto-population feature is. I would replace it with something like "If you just provide a user ID, it will attempt to auto-fill the other information based on existing user data." for clarity.

File: site/app/controllers/admin/UsersController.php [hunk 1]
```diff
@@ -943,30 +954,60 @@ public function uploadUserList($list_type = "classlist") {
...
+            // Remove the rows in which wrong or incorrect user_id is passed (only for row containing user_id)
+            if (count($row) == 1 && is_null($this->core->getQueries()->getSubmittyUser($row[0]))) {
+                $users_not_found[] = $row[0];
+                unset($users_to_add[$key]);
             }
-            if (isset($row[$pref_lastname_idx]) && !empty($row[$pref_lastname_idx])) {
-                $user->setPreferredLastName($row[$pref_lastname_idx]);
+        }
+        if (!empty($users_not_found)) {
+            $this->core->addErrorMessage('User(s) with following Username are not found ' . implode(', ', $users_not_found));
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]:
```suggestion
            $this->core->addErrorMessage('User(s) with following usernames are not found: ' . implode(', ', $users_not_found));
```

File: site/app/controllers/admin/UsersController.php [hunk 2]
```diff
@@ -923,9 +931,12 @@ public function uploadUserList($list_type = "classlist") {
             $exists = false;
             foreach ($existing_users as $i => $existing_user) {
                 if ($row[0] === $existing_user->getId()) {
-                    // Validate if this user has any data to update.
-                    // Did student registration section or grader group change?
-                    if ($row[4] !== $get_user_registration_or_group_function($existing_user)) {
+                    if (count($row) == 1) {
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]: Please always use `===` or `!==` for testing equality or inequality.

File: site/app/controllers/admin/UsersController.php [hunk 3]
```diff
@@ -943,30 +954,60 @@ public function uploadUserList($list_type = "classlist") {
...
-            $user->setLegalLastName($row[2]);
-            $user->setEmail($row[3]);
-            $set_user_registration_or_group_function($user);
-            if (isset($row[$pref_firstname_idx]) && !empty($row[$pref_firstname_idx])) {
-                $user->setPreferredFirstName($row[$pref_firstname_idx]);
+        $users_not_found = []; // track wrong user_ids given
+
+        foreach ($users_to_add as $key => $row) {
+            // Remove the rows in which wrong or incorrect user_id is passed (only for row containing user_id)
+            if (count($row) == 1 && is_null($this->core->getQueries()->getSubmittyUser($row[0]))) {
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]: Please always use `===` or `!==` for testing equality or inequality.

File: site/app/controllers/admin/UsersController.php [hunk 4]
```diff
@@ -943,30 +954,60 @@ public function uploadUserList($list_type = "classlist") {
...
-            if (isset($row[$pref_lastname_idx]) && !empty($row[$pref_lastname_idx])) {
-                $user->setPreferredLastName($row[$pref_lastname_idx]);
+        }
+        if (!empty($users_not_found)) {
+            $this->core->addErrorMessage('User(s) with following Username are not found ' . implode(', ', $users_not_found));
+            $this->core->redirect($return_url);
+        }
+        foreach ($users_to_add as $row) {
+            // Auto-populate user detail if only user_id is given
+            if (count($row) == 1) {
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]: Please always use `===` or `!==` for testing equality or inequality.

File: site/app/controllers/admin/UsersController.php [hunk 5]
```diff
@@ -943,30 +954,60 @@ public function uploadUserList($list_type = "classlist") {
...
+                $user = $this->core->getQueries()->getUserById($row[0]);
+                // set group as 'student' if upload is meant for classlist else set 'limited_access_grader' level
+                $user_group = $list_type == 'classlist' ? '4' : '3';
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]: Please always use `===` or `!==` for testing equality or inequality.

File: site/app/controllers/admin/UsersController.php [hunk 6]
```diff
@@ -943,30 +954,60 @@ public function uploadUserList($list_type = "classlist") {
...
             }
-            $insert_or_update_user_function('insert', $user);
         }
 
         // Existing users update
         foreach ($users_to_update as $row) {
             $user = $this->core->getQueries()->getUserById($row[0]);
             //Update registration section (student) or group (grader)
-            $set_user_registration_or_group_function($user);
+            if (count($row) == 1) {
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]: Please always use `===` or `!==` for testing equality or inequality.

File: site/app/controllers/admin/UsersController.php [hunk 7]
```diff
@@ -943,30 +954,60 @@ public function uploadUserList($list_type = "classlist") {
...
+                // set group as 'student' if upload is meant for classlist else set 'limited_access_grader' level
+                $user_group = $list_type == 'classlist' ? '4' : '3';
```
Comment (MasterOdin) [inline site/app/controllers/admin/UsersController.php]: Please always use `===` or `!==` for testing equality or inequality.

File: site/app/templates/admin/users/GraderListForm.twig [hunk 8]
```diff
@@ -3,8 +3,9 @@
 {% block title %}Upload Grader List{% endblock %}
 {% block body %}
     <p>
-        Format your grader data as an .xlsx or .csv file with {{ use_database ? 8 : 7 }} columns:<br>
+        Format your grader data as an .xlsx or .csv file with {{ use_database ? 8 : 7 }} columns :<br>
         &emsp;username, first name, last name, email, grader group, {{ use_database ? "password, " : "" }}preferred first name, preferred last name<br>
```
Comment (MasterOdin) [inline site/app/templates/admin/users/GraderListForm.twig]: While we're here, can you modify this to be a numbered list like the student upload form?

File: site/app/controllers/admin/UsersController.php (hunk 1 above)
Comment (mukul-kmr-jha) [reply site/app/controllers/admin/UsersController.php]: Done.

File: site/app/controllers/admin/UsersController.php (hunk 2 above)
Comment (mukul-kmr-jha) [reply site/app/controllers/admin/UsersController.php]: Done, thanks!

File: site/app/templates/admin/users/GraderListForm.twig (hunk 8 above)
Comment (mukul-kmr-jha) [reply site/app/templates/admin/users/GraderListForm.twig]: Done

**Comment by mukul-kmr-jha** [comment]
After reviewing this PR can you please take a look at this related PR #5133 . Thanks!

**Comment by MasterOdin** [comment]
Sorry it took so long to review this and merge this. Thanks!

**Pull request merged**